yarn-project/aztec-nr/aztec/src/state_vars/set.nr

```diff
@@ -76,42 +72,40 @@ impl<Note, N> Set<Note, N> {
     }
 
     // docs:start:remove
-    pub fn remove(self, note: Note) {
+    pub fn remove(self, note: Note) where Note: NoteInterface {
         let context = self.context.private.unwrap();
-        let note_hash = compute_note_hash_for_read_or_nullify(self.note_interface, note);
+        let note_hash = compute_note_hash_for_read_or_nullify(note);
         let has_been_read = context.read_requests.any(|r: SideEffect| r.value == note_hash);
         assert(has_been_read, "Can only remove a note that has been read from the set.");
 
         destroy_note(
             context,
             note,
-            self.note_interface,
         );
     }
     // docs:end:remove
 
     // docs:start:get_notes
-    pub fn get_notes<FILTER_ARGS>(
+    pub fn get_notes<N, FILTER_ARGS>(
         self,
         options: NoteGetterOptions<Note, N, FILTER_ARGS>,
-    ) -> [Option<Note>; MAX_READ_REQUESTS_PER_CALL] {
+    ) -> [Option<Note>; MAX_READ_REQUESTS_PER_CALL] where Note: NoteInterface + Serialize<N> + Deserialize<N> {
         let storage_slot = self.storage_slot;
         let opt_notes = get_notes(
             self.context.private.unwrap(),
             storage_slot,
-            self.note_interface,
             options,
         );
         opt_notes
     }
     // docs:end:get_notes
 
     // docs:start:view_notes
-    unconstrained pub fn view_notes(
+    unconstrained pub fn view_notes<N>(
         self,
         options: NoteViewerOptions<Note, N>,
-    ) -> [Option<Note>; MAX_NOTES_PER_PAGE] {
-        view_notes(self.storage_slot, self.note_interface, options)
+    ) -> [Option<Note>; MAX_NOTES_PER_PAGE] where Note: NoteInterface + Deserialize<N> {
```

Comment:
This syntax is incredible.

Reply:
And in the future we might be able to put trait constraints in the impl block so we don't have to repeat ourselves on every fn of the impl :D